aten/src/ATen/native/cuda/TensorShape.cu

```diff
@@ -0,0 +1,385 @@
+
+  // Calculate the number of blocks required for the first chunk across all
+  // splits, assuming each thread only processes BYTES_PER_THREAD bytes.
+  int64_t num_blocks = 0;
+  for (const auto& split_chunk_size : split_chunk_sizes) {
+    num_blocks +=
+        detail::div_up(split_chunk_size, BLOCK_SIZE * BYTES_PER_THREAD);
+  }
+
+  // Calculate the maximum number of blocks to launch. Only consider
+  // maxThreadsPerMultiProcessor as a limiting factor as the kernel uses no
+  // shared memory and little registers. Over-subscribe the SMs to hide I/O
+  // latency.
+  const auto num_sms =
+      at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
+  const auto max_threads_per_sm =
+      at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor;
+  const int64_t max_blocks = num_sms * max_threads_per_sm / BLOCK_SIZE * 2.0;
+
+  // Make each thread process BYTES_PER_THREAD * iter_factor bytes to regulate
+  // block size. Spread iter_factor evenly between chunks_per_block and
+  // iters_per_chunk.
+  int64_t iter_factor = detail::div_up(num_blocks * num_chunks, max_blocks);
+  int64_t chunks_per_block = std::ceil(std::sqrt(iter_factor));
+  chunks_per_block = std::min(chunks_per_block, num_chunks);
+  const int64_t iters_per_chunk = detail::div_up(iter_factor, chunks_per_block);
+
+  // Launch a logically jagged grid of shape
+  // (chunk_size*, num_splits, num_chunks / chunks_per_block)
+  // backed by a physical grid of shape
+  // (sum(chunk_size), num_chunks / chunks_per_block).
+  // A block can find its split_idx via block_idx_to_split_idx.
+  std::vector<int64_t> block_idx_to_split_idx;
+  std::vector<int64_t> blocks_cumsums{0};
+  block_idx_to_split_idx.reserve(num_blocks);
+  for (size_t split_idx = 0; split_idx < split_sizes.size(); ++split_idx) {
+    const auto blocks = detail::div_up(
+        split_chunk_sizes[split_idx],
+        BLOCK_SIZE * BYTES_PER_THREAD * iters_per_chunk);
+    block_idx_to_split_idx.insert(
+        block_idx_to_split_idx.end(), blocks, split_idx);
+    blocks_cumsums.push_back(blocks_cumsums.back() + blocks);
+  }
+
+  dim3 blocks(blocks_cumsums.back(), num_chunks / chunks_per_block, 1);
+  dim3 threads(BLOCK_SIZE, 1, 1);
+
+  auto [_, ptrs] = detail::pack_vecs(
+      {&dst_base_addrs,
+       &src_base_addrs,
+       &split_chunk_sizes,
+       &block_idx_to_split_idx,
+       &blocks_cumsums},
+      device);
+
+  detail::split_with_sizes_copy_out_contiguous_no_cast_kernel<<<
+      blocks,
+      threads,
+      0,
+      at::cuda::getCurrentCUDAStream()>>>(
+      /*dst_base_addrs=*/reinterpret_cast<char**>(ptrs[0]),
+      /*src_base_addrs=*/reinterpret_cast<char**>(ptrs[1]),
+      /*split_chunk_sizes=*/ptrs[2],
+      /*block_idx_to_split_idx=*/ptrs[3],
+      /*blocks_cumsums=*/ptrs[4],
+      src_stride,
+      num_chunks);
+  C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+  for (const auto& t : out) {
+    t.unsafeGetTensorImpl()->bump_version();
```

Comment:
This should be already done by the kernel generated for the ADInplaceOrView key by codegen?